File: app/src/main/java/com/dpis/module/templates/QuickTemplateApplyCoordinator.java
package com.dpis.module.templates;

import java.util.ArrayList;
import java.util.Collections;
import java.util.LinkedHashSet;
import java.util.List;
import java.util.Set;

/**
 * Coordinates applying a saved quick-template value to selected packages.
 *
 * <p>The coordinator is intentionally generic so template apply semantics can
 * live in the template package without knowing root-owned package config,
 * runtime publishing, or app-list models.</p>
 */
@SuppressWarnings("java:S1845")
public final class QuickTemplateApplyCoordinator<T> {
    public interface Template<T> {
        Set<String> selectedPackages();

        T configValue();
    }

    public interface ConfigWriter<T> {
        boolean hasRealPackageConfig(String packageName);

        boolean writePackageTemplateConfigValue(String packageName, T value);
    }

    public interface RuntimePublisher<T> {
        void publish(String packageName, T value);
    }

    public interface TargetPackageFilter {
        boolean isAllowed(String packageName);
    }

    private final ConfigWriter<T> configWriter;
    private final RuntimePublisher<T> runtimePublisher;

    public QuickTemplateApplyCoordinator(
            ConfigWriter<T> configWriter,
            RuntimePublisher<T> runtimePublisher) {
        this.configWriter = configWriter;
        this.runtimePublisher = runtimePublisher;
    }

    public Plan plan(Template<T> template) {
        return plan(template, null);
    }

    public Plan plan(Template<T> template, TargetPackageFilter targetPackageFilter) {
        LinkedHashSet<String> targets = sanitizePackages(
                template != null ? template.selectedPackages() : null);
        targets = filterPackages(targets, targetPackageFilter);
        int overwriteCount = 0;
        if (configWriter != null) {
            for (String packageName : targets) {
                if (configWriter.hasRealPackageConfig(packageName)) {
                    overwriteCount++;
                }
            }
        }
        return new Plan(targets.size(), overwriteCount);
    }

    public Result apply(Template<T> template) {
        return apply(template, null);
    }

    public Result apply(Template<T> template, TargetPackageFilter targetPackageFilter) {
        if (configWriter == null || template == null) {
            return Result.failure(Collections.emptyList(), Collections.emptyList());
        }
        LinkedHashSet<String> targets = sanitizePackages(template.selectedPackages());
        targets = filterPackages(targets, targetPackageFilter);
        if (targets.isEmpty()) {
            return Result.noSelection();
        }
        ArrayList<String> successfulPackages = new ArrayList<>();
        ArrayList<String> failedPackages = new ArrayList<>();
        T configValue = template.configValue();
        for (String packageName : targets) {
            boolean saved = configWriter.writePackageTemplateConfigValue(packageName, configValue);
            if (saved) {
                successfulPackages.add(packageName);
                if (runtimePublisher != null) {
                    runtimePublisher.publish(packageName, configValue);
                }
            } else {
                failedPackages.add(packageName);
            }
        }
        return new Result(false, successfulPackages, failedPackages);
    }

    private static LinkedHashSet<String> sanitizePackages(Set<String> packageNames) {
        LinkedHashSet<String> sanitized = new LinkedHashSet<>();
        if (packageNames == null) {
            return sanitized;
        }
        for (String packageName : packageNames) {
            if (packageName == null) {
                continue;
            }
            String trimmed = packageName.trim();
            if (!trimmed.isEmpty()) {
                sanitized.add(trimmed);
            }
        }
        return sanitized;
    }

    private static LinkedHashSet<String> filterPackages(
            LinkedHashSet<String> packageNames,
            TargetPackageFilter targetPackageFilter) {
        if (targetPackageFilter == null || packageNames == null || packageNames.isEmpty()) {
            return packageNames != null ? packageNames : new LinkedHashSet<>();
        }
        LinkedHashSet<String> filtered = new LinkedHashSet<>();
        for (String packageName : packageNames) {
            if (targetPackageFilter.isAllowed(packageName)) {
                filtered.add(packageName);
            }
        }
        return filtered;
    }

    public static final class Plan {
        public final int targetCount;
        public final int overwriteCount;

        private Plan(int targetCount, int overwriteCount) {
            this.targetCount = targetCount;
            this.overwriteCount = overwriteCount;
        }
    }

    public static final class Result {
        public final boolean emptySelection;
        public final List<String> successfulPackages;
        public final List<String> failedPackages;

        private Result(boolean emptySelection,
                List<String> successfulPackages,
                List<String> failedPackages) {
            this.emptySelection = emptySelection;
            this.successfulPackages = Collections.unmodifiableList(new ArrayList<>(
                    successfulPackages != null ? successfulPackages : Collections.emptyList()));
            this.failedPackages = Collections.unmodifiableList(new ArrayList<>(
                    failedPackages != null ? failedPackages : Collections.emptyList()));
        }

        public static Result noSelection() {
            return new Result(true, Collections.emptyList(), Collections.emptyList());
        }

        public static Result failure(List<String> successfulPackages, List<String> failedPackages) {
            return new Result(false, successfulPackages, failedPackages);
        }

        public int successCount() {
            return successfulPackages.size();
        }

        public int failureCount() {
            return failedPackages.size();
        }
    }
}

File: app/src/main/java/com/dpis/module/templates/QuickTemplateApplyCoordinator.kt
package com.dpis.module.templates

import java.util.Collections

/**
 * Coordinates applying a saved quick-template value to selected packages.
 * 
 * 
 * The coordinator is intentionally generic so template apply semantics can
 * live in the template package without knowing root-owned package config,
 * runtime publishing, or app-list models.
 */
class QuickTemplateApplyCoordinator<T>(
    private val configWriter: ConfigWriter<T?>?,
    private val runtimePublisher: RuntimePublisher<T?>?
) {
    interface Template<T> {
        fun selectedPackages(): MutableSet<String?>?

        fun configValue(): T?
    }

    interface ConfigWriter<T> {
        fun hasRealPackageConfig(packageName: String?): Boolean

        fun writePackageTemplateConfigValue(packageName: String?, value: T?): Boolean
    }

    interface RuntimePublisher<T> {
        fun publish(packageName: String?, value: T?)
    }

    fun interface TargetPackageFilter {
        fun isAllowed(packageName: String?): Boolean
    }

    @JvmOverloads
    fun plan(template: Template<T?>?, targetPackageFilter: TargetPackageFilter? = null): Plan {
        var targets: LinkedHashSet<String?> = sanitizePackages(
            if (template != null) template.selectedPackages() else null
        )
        targets = filterPackages(targets, targetPackageFilter)
        var overwriteCount = 0
        if (configWriter != null) {
            for (packageName in targets) {
                if (configWriter.hasRealPackageConfig(packageName)) {
                    overwriteCount++
                }
            }
        }
        return Plan(targets.size, overwriteCount)
    }

    @JvmOverloads
    fun apply(template: Template<T?>?, targetPackageFilter: TargetPackageFilter? = null): Result {
        if (configWriter == null || template == null) {
            return Result.Companion.failure(mutableListOf<String?>(), mutableListOf<String?>())
        }
        var targets: LinkedHashSet<String?> = sanitizePackages(template.selectedPackages())
        targets = filterPackages(targets, targetPackageFilter)
        if (targets.isEmpty()) {
            return Result.Companion.noSelection()
        }
        val successfulPackages = ArrayList<String?>()
        val failedPackages = ArrayList<String?>()
        val configValue = template.configValue()
        for (packageName in targets) {
            val saved = configWriter.writePackageTemplateConfigValue(packageName, configValue)
            if (saved) {
                successfulPackages.add(packageName)
                if (runtimePublisher != null) {
                    runtimePublisher.publish(packageName, configValue)
                }
            } else {
                failedPackages.add(packageName)
            }
        }
        return Result(false, successfulPackages, failedPackages)
    }

    class Plan(@JvmField val targetCount: Int, @JvmField val overwriteCount: Int)

    class Result(
        @JvmField val emptySelection: Boolean,
        successfulPackages: MutableList<String?>?,
        failedPackages: MutableList<String?>?
    ) {
        @JvmField
        val successfulPackages: MutableList<String?>
        @JvmField
        val failedPackages: MutableList<String?>

        init {
            this.successfulPackages = Collections.unmodifiableList<String?>(
                ArrayList<String?>(
                    if (successfulPackages != null) successfulPackages else mutableListOf<String?>()
                )
            )
            this.failedPackages = Collections.unmodifiableList<String?>(
                ArrayList<String?>(
                    if (failedPackages != null) failedPackages else mutableListOf<String?>()
                )
            )
        }

        fun successCount(): Int {
            return successfulPackages.size
        }

        fun failureCount(): Int {
            return failedPackages.size
        }

        companion object {
            fun noSelection(): Result {
                return Result(true, mutableListOf<String?>(), mutableListOf<String?>())
            }

            fun failure(
                successfulPackages: MutableList<String?>?,
                failedPackages: MutableList<String?>?
            ): Result {
                return Result(false, successfulPackages, failedPackages)
            }
        }
    }

    companion object {
        private fun sanitizePackages(packageNames: MutableSet<String?>?): LinkedHashSet<String?> {
            val sanitized = LinkedHashSet<String?>()
            if (packageNames == null) {
                return sanitized
            }
            for (packageName in packageNames) {
                if (packageName == null) {
                    continue
                }
                val trimmed = packageName.trim { it <= ' ' }
                if (!trimmed.isEmpty()) {
                    sanitized.add(trimmed)
                }
            }
            return sanitized
        }

        private fun filterPackages(
            packageNames: LinkedHashSet<String?>?,
            targetPackageFilter: TargetPackageFilter?
        ): LinkedHashSet<String?> {
            if (targetPackageFilter == null || packageNames == null || packageNames.isEmpty()) {
                return if (packageNames != null) packageNames else LinkedHashSet<String?>()
            }
            val filtered = LinkedHashSet<String?>()
            for (packageName in packageNames) {
                if (targetPackageFilter.isAllowed(packageName)) {
                    filtered.add(packageName)
                }
            }
            return filtered
        }
    }
}
